ethcore/src/client/client.rs

```diff
@@ -2583,6 +2582,100 @@ impl ProvingBlockChainClient for Client {
 
 impl SnapshotClient for Client {}
 
+impl ExportBlocks for Client {
+	fn export_blocks<'a>(&self, mut out: Box<dyn std::io::Write + 'a>, from: BlockId, to: BlockId, format: Option<DataFormat>) -> Result<(), String> {
+		let from = self.block_number(from).ok_or("From block could not be found")?;
+		let to = self.block_number(to).ok_or("To block could not be found")?;
+		let format = format.unwrap_or_default();
+
+		for i in from..(to + 1) {
+			if i % 10000 == 0 {
+				info!("#{}", i);
+			}
+			let b = self.block(BlockId::Number(i)).ok_or("Error exporting incomplete chain")?.into_inner();
+			match format {
+				DataFormat::Binary => {
+					out.write(&b).map_err(|e| format!("Couldn't write to stream. Cause: {}", e))?;
+				}
+				DataFormat::Hex => {
+					out.write_fmt(format_args!("{}\n", b.pretty())).map_err(|e| format!("Couldn't write to stream. Cause: {}", e))?;
+				}
+			}
+		}
+		Ok(())
+	}
+}
+
+
+impl ImportBlocks for Client {
+	fn import_blocks<'a>(&self, mut source: Box<dyn std::io::Read + 'a>, format: Option<DataFormat>) -> Result<(), String> {
+		const READAHEAD_BYTES: usize = 8;
+
+		let mut first_bytes: Vec<u8> = vec![0; READAHEAD_BYTES];
+		let mut first_read = 0;
+
+		let format = match format {
+			Some(format) => format,
+			None => {
+				first_read = source.read(&mut first_bytes).map_err(|_| "Error reading from the file/stream.")?;
+				match first_bytes[0] {
+					0xf9 => DataFormat::Binary,
+					_ => DataFormat::Hex,
+				}
+			}
+		};
+
+		let do_import = |bytes| {
+			let block = Unverified::from_rlp(bytes).map_err(|_| "Invalid block rlp")?;
+			while self.queue_info().is_full() { std::thread::sleep(Duration::from_secs(1)); }
+			match self.import_block(block) {
+				Err(EthcoreError::Import(ImportError::AlreadyInChain)) => {
+					trace!("Skipping block already in chain.");
+				}
+				Err(e) => {
+					return Err(format!("Cannot import block: {:?}", e));
+				},
+				Ok(_) => {},
+			}
+			Ok(())
+		};
+
+		match format {
+			DataFormat::Binary => {
+				loop {
+					let mut bytes = if first_read > 0 {
+						first_bytes.clone()
+					} else {
+						vec![0; READAHEAD_BYTES]
+					};
+					let n = if first_read > 0 {
+						first_read
+					} else {
+						source.read(&mut bytes).map_err(|_| "Error reading from the file/stream.")?
```

Comment:
The specific error should be displayed, for ease of debugging.  Also, this code does not handle partial reads.  On \*nix, this means that it works for reading regular files, but not pipes or sockets.  I do not know about Windows.

Reply:
It works perfectly for pipes, sockets aren't particularly useful in this case. Also not sure what you mean by

>  Also, this code does not handle partial read.

Reply:
So I'm trying to understand, from what you're saying `read` can return 0 bytes even though there are still bytes to be read?

Sockets are not particularly useful in this case because you can't exactly back up your blockchain by writing it to a socket, or attempt to import chain data out of a socket.

Reply:
On partial reads - https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact

Sockets are not particularly useful in this case because it's not exactly _feasible_ back up your blockchain by writing it to a socket or import chain data out of a socket.